### PR TITLE
[PLUGIN-1715] Implement retry feature for BigQuery Execute Plugin

### PR DESCRIPTION
**Pull request by psainics**
## Implement retry feature for BigQuery Execute Plugin 

Jira : [PLUGIN-1715](https://cdap.atlassian.net/browse/PLUGIN-1715)

- In general internal errors should be retried.
- Do an exponential back-off retry on errors with reason `jobBackendError` , `jobInternalError`.
- Backoff requirements (exponential from 1 to 32 seconds with a multiplier of 2)

### Code Changes
- Add a new set that contains all reason we should retry on.
- A new custom exception that is raised when the error should be retried.
- A new function to generate the exponential back- off retry policy  
- A new dependency is added `dev.failsafe` that handles the retry logic based on the exception raised.

### Unit Tests
- Use mocks to test the behavior of the retry.

[PLUGIN-1715]: https://cdap.atlassian.net/browse/PLUGIN-1715?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java]: Don't use jetbrains annotations in production code. This may bring unnessesary dependencies

File: src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java]: Usually package-only access is enough for testing

Comment (tivv) [reply src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java]: Adding an explicit config would effectively remove a no-argument constructor. Which constructor will be used for the production run? If this one, it should not be marked as TestOnly

File: src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java]: All constant should be uppercased

File: src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java]: All of this should be configurable. Retries must be disableable

File: src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java:212]: It's a checked exception. What would it be converted to when we are out of retries?

File: src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java
Comment (tivv) [inline src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: You probably want immediate retries, otherwise this test would be waiting unnesessary

File: src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java
Comment (tivv) [inline src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: Why are you expecting just a RuntimeException here? It that it should go through retries and give some reasonable error

File: src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java]: If you want to mark a method as visible only for the purposes of testing, use `@VisibleForTesting` which is already used elsewhere in the codebase.

File: src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java]: Also the strings here should be in their own constants.

File: src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java
Comment (fernst) [reply src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java]: Agree. Please add a `Retry Configuration` section in the UI and allow the users to fine-tune this behavior.

**Review by fernst — COMMENTED** [review]
Agree with Vitalii's suggestions here. Please address.

File: src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java
Comment (vikasrathee-cs) [reply src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java]: @tivv @fernst It was mentioned previously on the bug ticket also that Bigquery SLA is for 32 seconds and that is the reason we have added these fix properties to go upto that limit. Same was confirmed in design doc also.
Can you please confirm in design doc also. Shared with you.
Also have a look at this. https://cloud.google.com/bigquery/sla#:~:text=%22-,Back%2Doff%20Requirements,-%22%C2%A0means%2C%20when

Comment (tivv) [reply src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java]: Few points here:

-  Customer may not wish to wait so long. Yep, BQ may recover, but it's also possible that it's too long for given customer
- You are changing the behaviour. We must have an option to disable this behavior
- SLAs are statistical. They may still be violated
- SLAs are point in time. They may change and we need to be able to adjust the bahvior without code change

File: src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java
Comment (psainics) [reply src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java]: Updated !

File: src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java
Comment (psainics) [reply src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java]: Removed the jetbrains annotation,  using `@ VisibleForTesting`

File: src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java
Comment (psainics) [reply src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java]: I have added new UI field to turn this behavior off (default) and let user configure the parameters.

image

File: src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java
Comment (psainics) [reply src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java]: Made this package only.

File: src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java
Comment (tivv) [inline src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java]: This way you are loosing all the additional information from original BigQueryException stack trace. There are only some case where you would want to hide all details of the underlying exception by using e.getMessage instead of attaching a root cause. I don't see why would we want to have it here. Same applies to the catch clause in `executeQuery`

Comment (psainics) [reply src/main/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecute.java]: Updated !

File: src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java
Comment (psainics) [reply src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: Fixed !

File: src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java
Comment (psainics) [reply src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: Updated !

**Comment by psainics** [comment]
Changes Squashed !

File: src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java
Comment (tivv) [inline src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: I am still a bit confused what final sequence of exceptions would it be and, what's important, what would be a message to user. Could you please use assertThrows and validate the message as described here: https://www.baeldung.com/junit-assert-exception ? I am wondering if we would need to unwrap those exceptions to prevent unneeded layers / cluttered message

Comment (psainics) [reply src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: The user will always see a `BigQueryJobExecutionException`

There are 3 cases for this to occur
1. Retry is not possible as the error reason is not in reason list, the user get the message `Failed to execute BigQuery job. Reason: {Reason from that exception}`

2. Retry was tried but failed , the user will get the message ~`Failed to execute BigQuery job.  io.cdap.plugin.gcp.bigquery.exception.BigQueryJobExecutionException`~ `Failed to execute BigQuery job. Reason: Retries exhausted.`
- We could provide a different message here if required, as the exception class will be  `FailsafeException` for which an additional catch could be added .

3. Someting else fails and another exception is thrown , `Failed to execute BigQuery job. Reason: {Reason from that exception}`

Comment (tivv) [reply src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: I am curious about the text of the message. I want to avoid something like  `Failed to execute BigQuery job. Reason: RuntimeException: BigQueryJobExecutionException: reason`

Comment (psainics) [reply src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: I have updated to assert throw, and made it more clear what will the end user see. in all cases.

**Review by fernst — APPROVED** [review]
LGTM

File: src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java
Comment (tivv) [inline src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: So, basically we say "Failed to execute BigQuery job. Reason: Job execution failed with error: $error" This is what I want to avoid as it's saying same thing twice

Comment (tivv) [reply src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: May be in top level catch we should just find the correct cause and rethrow in instead of wrapping it again

Comment (psainics) [reply src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: Sure, i have removed the wrapping of error message.

Comment (tivv) [reply src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: But this way you are loosing the underlying error. You should either: 
 - do visa versa and pop up only the underlying error message "Job execution failed with error: $error"
  - prepend it with some text that adds value, e.g. "Job retries failed after X tries. Last try failure: Job execution failed with error: $error".

Comment (psainics) [reply src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: Yup I am doing the 1st one , by removing wrapping I ment removing the extra messages string that was present before

Comment (tivv) [reply src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: But you are hiding the underlying "$error", so the only thing that we can see is that retries exhausted without much information about the underlying error.

Comment (psainics) [reply src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: As i already have logging when nth retry is made or the policy is out of retry should i just pop the error and throw that ?

something like this
```java
try {
      Failsafe.with(getRetryPolicy()).run(() -> executeQuery(bigQuery, queryConfig, context));
    } catch (FailsafeException e) {
      throw e.getCause();
    } 
``` 
Also i could drop the `catch (RuntimeException e)`

Comment (tivv) [reply src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: Sure, that's option #1 in the list above

Comment (tivv) [reply src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: I would check if e.getCause is not null as generally not every failure is an exception

**Comment by psainics** [comment]
Thanks @tivv , i have made the changes and squashed them , hopefully you can approve this now.
I will get this QA tested before merging  👍

File: src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java
Comment (psainics) [reply src/test/java/io/cdap/plugin/gcp/bigquery/action/BigQueryExecuteTest.java]: 🤔 
```java
if (e.getCause() != null) {
        throw e.getCause();  
} 
  throw e;
```

File: widgets/BigQueryExecute-action.json
Comment (itsankit-google) [inline widgets/BigQueryExecute-action.json]: shouldn't this be set to `true` by default, @tivv @fernst wdyt??
I thought aim is to add some resilency to the plugin to reduce intermittent failures but if the behavior is `false` by default, users will need to enable it explicitly and modify every pipeline.

Comment (psainics) [reply widgets/BigQueryExecute-action.json]: Your point makes a lot of sense. Setting this behavior to true by default aligns with the goal of enhancing plugin resilience and reducing intermittent failures.

Comment (tivv) [reply widgets/BigQueryExecute-action.json]: So, we usually have two things:
 * Default for UI. This is what we want default experience to be. In this case it should be "true"
 * Default for backend (when option is not set). This is usually how things worked before to ensure backward compatibility unless we have explicit OK from product to break backward compatibility

Comment (psainics) [reply widgets/BigQueryExecute-action.json]: @itsankit-google we had a discussion with deepinder, the UI fields should be hidden and true by default !

**Comment by psainics** [comment]
New changes include 
- Reverted changes in `docs/BigQueryExecute-action.md` as the UI fields are hidden.
- Removed filter `retryOnBackendError` with condition `retryOnBackendError == true` as this will not be required now.
- Made all the ui fields `hidden`
- Retry On Backend Error is `TRUE` by default , on widget and backed.

**Comment by psainics** [comment]
Resolved conflicts caused by PR: https://github.com/data-integrations/google-cloud/pull/1344 in `bigquery/action/BigQueryExecute.java`

**Review by tivv — APPROVED** [review]

**Pull request merged**
